libs/ml/benchmark/optimisation/optimisation.cpp

```diff
@@ -0,0 +1,285 @@
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+BENCHMARK_TEMPLATE(
+    BM_Optimiser_Construct,
+    fetch::ml::optimisers::SGDOptimiser<fetch::math::Tensor<fetch::fixed_point::fp128_t>>)
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+
+BENCHMARK_TEMPLATE(BM_Optimiser_Construct,
+                   fetch::ml::optimisers::AdamOptimiser<fetch::math::Tensor<float>>)
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+BENCHMARK_TEMPLATE(BM_Optimiser_Construct,
+                   fetch::ml::optimisers::AdamOptimiser<fetch::math::Tensor<double>>)
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+BENCHMARK_TEMPLATE(
+    BM_Optimiser_Construct,
+    fetch::ml::optimisers::AdamOptimiser<fetch::math::Tensor<fetch::fixed_point::fp32_t>>)
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+BENCHMARK_TEMPLATE(
+    BM_Optimiser_Construct,
+    fetch::ml::optimisers::AdamOptimiser<fetch::math::Tensor<fetch::fixed_point::fp64_t>>)
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+BENCHMARK_TEMPLATE(
+    BM_Optimiser_Construct,
+    fetch::ml::optimisers::AdamOptimiser<fetch::math::Tensor<fetch::fixed_point::fp128_t>>)
+    ->Apply(OptimiserChargeArguments)
+    ->Unit(::benchmark::kMicrosecond);
+
+template <class Optimiser>
+void BM_Optimiser_Run(benchmark::State &state)
+{
+  using TensorType = typename Optimiser::TensorType;
+  using DataType   = typename Optimiser::DataType;
+
+  fetch::SetGlobalLogLevel(fetch::LogLevel::ERROR);
+
+  // Get args from state
+  BM_Tensor_config config{state};
+
+  SizeType batch_size  = config.shape[0];
+  SizeType input_size  = config.shape[1];
+  SizeType hidden_size = config.shape[2];
+  SizeType output_size = config.shape[3];
+  SizeType n_epochs    = config.shape[4];
+  SizeType n_hidden    = config.shape[5];
+
+  auto learning_rate = fetch::math::Type<DataType>("0.001");
+
+  // Prepare data and labels
+  TensorType data({input_size, batch_size});
+  TensorType gt({output_size, batch_size});
+  data.FillUniformRandom();
+  gt.FillUniformRandom();
+
+  // make a graph
+  std::string input_name, label_name, error_name;
+  auto        g = make_graph<TensorType>(input_size, n_hidden, hidden_size, output_size, input_name,
+                                  label_name, error_name);
+
+  // Initialise Optimiser
+  Optimiser optimiser(g, {input_name}, label_name, error_name, learning_rate);
+
+  state.counters["charge"] = static_cast<double>(optimiser.ChargeStep() * n_epochs);
+
+  for (auto _ : state)
+  {
+    // Do optimisation
+    for (SizeType i = 0; i < n_epochs; ++i)
+    {
+      optimiser.Run({data}, gt);
+    }
+  }
+}
+
+static void OptimiserRunArguments(benchmark::internal::Benchmark *b)
+{
+  std::vector<std::int64_t> batch_size{1, 10, 100};
+
+  for (auto const &bs : batch_size)
+  {
+    b->Args({5, bs, bs, bs, bs, 100, 0});
```

Comment:
these seems confusing. batch size is set at position 1, 2, 3 & 4.
Looking at Optimiser run batch_size takes the config position 0, and uses this batch size for input, hidden, & output ?